README.md
# Générateur de certificat de déplacement

## Développer

### Installer le projet

```console
git clone https://github.com/LAB-MI/attestation-deplacement-derogatoire-covid-19.git
cd attestation-deplacement-derogatoire-covid-19
npm i
npm start
```

## Générer et tester le code de production

### Tester le code de production en local

#### Générer le code de production pour tester que le build fonctionne en entier

```console
npm run build:dev
```

#### Tester le code de production en local

```console
npx serve dist
```

Et visiter http://localhost:5000

Le code à déployer sera le contenu du dossier `dist`

## Docker Installation

#### Build de l'image

```console
git clone https://github.com/LAB-MI/attestation-deplacement-derogatoire-covid-19.git
cd attestation-deplacement-derogatoire-covid-19
docker build -t attestation-covid .
```
#### Démarrage du container via docker

``` console
docker run --name attestation-covid -p 5000:5000 attestation-covid:latest
```
#### Démarrage du container via docker-compose

``` console
docker-compose up -d
```

## Crédits

Ce projet a été réalisé à partir d'un fork du dépôt [deplacement-covid-19](https://github.com/nesk/deplacement-covid-19) de lui-même réalisé à partir d'un fork du dépôt [covid-19-certificate](https://github.com/nesk/covid-19-certificate) de [Johann Pardanaud](https://github.com/nesk).

Les projets open source suivants ont été utilisés pour le développement de ce
service :

- [PDF-LIB](https://pdf-lib.js.org/)
- [remove-accents](https://github.com/tyxla/remove-accents)
- [secure-ls](https://github.com/softvar/secure-ls)
- [qrcode](https://github.com/soldair/node-qrcode)
- [Bootstrap](https://getbootstrap.com/)
- [Font Awesome](https://fontawesome.com/license)

Pour une meilleure accessibilité des documents pdf générés, nous utilisons la police Luciole (https://luciole-vision.com/)


CHANGELOG.md
# Changelog

## [2.9.2]

- Correction du qr-code pour qu'il ne contienne que les motifs liés au contexte sélectionné

## [2.9.1]

- Suppression d'une virgule en trop dans quarantine.json (proposé par @Bbillyben)

## [2.9.0]

- Version générique pour l'application des arrêtés préfectoraux en vue de freiner l'épidémie
- Réactivation des onglets
- Adaptation des motifs
- Ajout d'un lien vers les sites internet des préfectures
  
## [2.8.1]

- Mise à jour du décret n°2021-699 du 1er juin 2021

## [2.8.0]

- Nouvelles dispositions applicables au 9 juin 2021

## [2.7.0]

- Nouvelles dispositions applicables au 19 mai 2021

## [2.6.0]

- Nouvelles dispositions applicables au 3 mai 2021
- Désactivation des onglets / motifs couvre-feu uniquement
- Positionnement vertical sur le PDF calculé

## [2.5.3]

- Correction warning console

## [2.5.2]

- Amélioration UI
- Corrections typographiques

- ## [2.5.1]

- Corrections typographiques

## [2.5.0]

- Modifications liées aux annonces gouvernementales du 31/03/2021
- Génération complète du PDF par l'application à partir d'un fichier de configuration JSON
- Amélioration de l'accessibilité : utilisation de la police LUCIOLE (<https://www.luciole-vision.com/>) conçue spécifiquement pour les personnes malvoyantes

## [2.4.1]

- Corrections typographiques

## [2.4.0]

- Modifications liées aux annonces gouvernementales du 18/03/2021
- Développement de la partie Mesures renforcées

## [2.3.0]

- Modifications liées aux annonces gouvernementales du 18/03/2021
- Développement de la partie Couvre-feu
- Préparation de la partie Mesures renforcées

## [2.2.0]

- Ajout de boutons pour choisir le contexte
- Correction de liens erronés, et informations manquantes

## [2.1.0]

- Correction de motif
- Reconstruction du formulaire
- Modification du pdf (certificate.pdf)
- Repositionnement des textes ajoutés

## [2.0.0]

- Reconstruction du formulaire
- Modification du pdf
- Repositionnement des textes ajoutés

## [1.1.0]

- Modification du pdf (certificate.pdf)
- Repositionnement des textes ajoutés
- Accessibilité : ajout d'un texte (en blanc) pour indiquer aux lecteurs de pdf audio la présence d'un QR-code dans le document généré

## [1.1.1]

- Modification du pdf (certificate.pdf)
- Repositionnement des textes ajoutés

## [1.1.2]

- Ajout des liens application TousAntiCovid


src/js/form.js
import 'bootstrap/dist/css/bootstrap.min.css'

import '../css/main.css'

import formData from '../form-data.json'

import { $, appendTo, createElement } from './dom-utils'

const createTitle = () => {
  const h2 = createElement('h2', { className: 'titre-2', innerHTML: 'Remplissez en ligne votre déclaration numérique : ' })
  const p = createElement('p', { className: 'msg-info', innerHTML: 'Tous les champs sont obligatoires.' })
  return [h2, p]
}

const createFormGroup = ({
  autocomplete = false,
  autofocus = false,
  inputmode,
  label,
  max,
  min,
  maxlength,
  minlength,
  name,
  pattern,
  placeholder = '',
  type = 'text',
}) => {
  const formGroup = createElement('div', { className: 'form-group' })
  const labelAttrs = {
    for: `field-${name}`,
    id: `field-${name}-label`,
    innerHTML: label,
  }
  const labelEl = createElement('label', labelAttrs)

  const inputGroup = createElement('div', { className: 'input-group align-items-center' })
  const inputAttrs = {
    autocomplete,
    autofocus,
    className: 'form-control',
    id: `field-${name}`,
    inputmode,
    min,
    max,
    minlength,
    maxlength,
    name,
    pattern,
    placeholder,
    required: true,
    type,
  }

  const input = createElement('input', inputAttrs)

  const validityAttrs = {
    className: 'validity',
  }
  const validity = createElement('span', validityAttrs)

  const example = createElement('p', { className: 'exemple  basis-100' })

  const appendToFormGroup = appendTo(formGroup)
  appendToFormGroup(labelEl)
  appendToFormGroup(inputGroup)

  const appendToInputGroup = appendTo(inputGroup)
  appendToInputGroup(input)
  appendToInputGroup(validity)
  appendToInputGroup(example)

  return formGroup
}

function createReasonField (context) {
  return (reasonData) => {
    const formReasonAttrs = { className: 'form-checkbox align-items-center' }
    const formReason = createElement('div', formReasonAttrs)
    const appendToReason = appendTo(formReason)

    const id = `checkbox-${context}-${reasonData.code}`
    const className = `form-check-input ${context}`
    const inputReasonAttrs = {
      className,
      type: 'checkbox',
      id,
      name: 'field-reason',
      value: reasonData.code,
    }
    const inputReason = createElement('input', inputReasonAttrs)

    if (reasonData.type === 'title') {
      const titleAttrs = { innerHTML: reasonData.label, className: 'group-reason' }
      const title = createElement('h5', titleAttrs)
      appendToReason([title])
    } else {
      const labelAttrs = { innerHTML: reasonData.label, className: 'form-checkbox-label', for: id }
      const label = createElement('label', labelAttrs)
      appendToReason([inputReason, label])
    }
    return formReason
  }
}

const createReasonFieldset = (reasonsData, sanitaryContextData) => {
  const fieldsetAttrs = {
    id: 'curfew-reason-fieldset',
    className: 'fieldset  reason-fieldset',
  }

  const fieldset = createElement('fieldset', fieldsetAttrs)
  const appendToFieldset = appendTo(fieldset)

  const legendAttrs = {
    className: 'legend titre-3',
    innerHTML: 'Je me déplace en période de couvre-feu pour l\'une des raisons suivantes :',
  }
  const legend = createElement('p', legendAttrs)

  const reasonsFields = reasonsData.items.map(createReasonField(reasonsData.key))

  appendToFieldset([legend, ...reasonsFields])
  return fieldset
}

const createReasonFieldsetQuarantine = (reasonsData) => {
  const fieldsetAttrs = {
    id: 'quarantine-reason-fieldset',
    className: 'fieldset  reason-fieldset',
  }

  const fieldset = createElement('fieldset', fieldsetAttrs)
  const appendToFieldset = appendTo(fieldset)
  const reasonsFields = reasonsData.items.map(createReasonField(reasonsData.key))

  // const footnoteAttrs = { id: 'footnote2', className: 'footnote', innerHTML: ' [1] Pour les personnes résidant aux frontières d’un département, une tolérance de 30 kilomètres au-delà du département est acceptée. ' }
  // const footnote = createElement('p', footnoteAttrs)

  const legendAttrs = {
    className: 'legend titre-3',
    innerHTML: 'Je me déplace en journée pour l\'une des raisons suivantes :',
  }
  const legend = createElement('p', legendAttrs)

  appendToFieldset([legend, ...reasonsFields])
  // Créer un form-checkbox par motif
  return fieldset
}

export function createForm () {
  const form = $('#form-profile')
  // Évite de recréer le formulaire s'il est déjà créé par react-snap (ou un autre outil de prerender)
  if (form.innerHTML !== '') {
    return
  }

  const appendToForm = appendTo(form)

  const formFirstPart = formData
    .flat(1)
    .filter(field => field.key !== 'curfew')
    .filter(field => field.key !== 'quarantine')
    .filter(field => !field.isHidden)
    .map((field,
      index) => {
      const formGroup = createFormGroup({
        autofocus: index === 0,
        ...field,
        name: field.key,
      })

      return formGroup
    })

  const reasonsDataCurfew = formData
    .flat(1)
    .find(field => field.key === 'curfew')

  const reasonsDataQuarantine = formData
    .flat(1)
    .find(field => field.key === 'quarantine')

  const reasonFieldsetCurfew = createReasonFieldset(reasonsDataCurfew)
  const reasonFieldsetQuarantine = createReasonFieldsetQuarantine(reasonsDataQuarantine)
  const curfewButton = createElement('button', { type: 'button', className: 'curfew-button  context-button  btn', innerHTML: '<i class="fa fa-moon inline-block mr-1"></i> Couvre-feu' })
  const quarantineButton = createElement('button', { type: 'button', className: 'quarantine-button  context-button  btn', innerHTML: '<i class="fa fa-sun inline-block mr-1"></i> Journée' })
  const buttonWrapper = createElement('div', { className: 'button-wrapper' })
  buttonWrapper.appendChild(curfewButton)
  buttonWrapper.appendChild(quarantineButton)
  const contextTitle = createElement('p', { className: 'context-title' })
  const contextTitleText = document.createTextNode('Choisissez un contexte')
  contextTitle.appendChild(contextTitleText)
  const contextSubtitle = createElement('p', { className: 'context-subtitle' })
  const contextWrapper = createElement('div', { className: 'context-wrapper' })
  contextWrapper.appendChild(contextTitle)
  contextWrapper.appendChild(contextSubtitle)
  contextWrapper.appendChild(buttonWrapper)
  const reasonFielsetWrapper = createElement('div', { className: 'fieldset-wrapper  hidden' })

  reasonFielsetWrapper.appendChild(reasonFieldsetCurfew)
  reasonFielsetWrapper.appendChild(reasonFieldsetQuarantine)
  appendToForm([...createTitle(), ...formFirstPart, contextWrapper, reasonFielsetWrapper])
}


src/js/form-util.js
import removeAccents from 'remove-accents'

import { $, $$, downloadBlob } from './dom-utils'
import { addSlash, getFormattedDate } from './util'
import { generatePdf } from './pdf-util'
import SecureLS from 'secure-ls'
let context = 'curfew'
const secureLS = new SecureLS({ encodingType: 'aes' })
const clearDataSnackbar = $('#snackbar-cleardata')
const storeDataInput = $('#field-storedata')
const conditions = {
  '#field-firstname': {
    length: 1,
  },
  '#field-lastname': {
    length: 1,
  },
  '#field-birthday': {
    pattern: /^([0][1-9]|[1-2][0-9]|30|31)\/([0][1-9]|10|11|12)\/(19[0-9][0-9]|20[0-1][0-9]|2020)/g,
  },
  '#field-address': {
    length: 1,
  },
  '#field-city': {
    length: 1,
  },
  '#field-zipcode': {
    pattern: /\d{5}/g,
  },
  '#field-datesortie': {
    pattern: /\d{4}-\d{2}-\d{2}/g,
  },
  '#field-heuresortie': {
    pattern: /\d{2}:\d{2}/g,
  },
}

function validateAriaFields () {
  return Object.keys(conditions)
    .map((field) => {
      const fieldData = conditions[field]
      const pattern = fieldData.pattern
      const length = fieldData.length
      const isInvalidPattern = pattern && !$(field).value.match(pattern)
      const isInvalidLength = length && !$(field).value.length

      const isInvalid = !!(isInvalidPattern || isInvalidLength)

      $(field).setAttribute('aria-invalid', isInvalid)
      if (isInvalid) {
        $(field).focus()
      }
      return isInvalid
    })
    .includes(true)
}

function updateSecureLS (formInputs) {
  if (wantDataToBeStored() === true) {
    secureLS.set('profile', getProfile(formInputs))
  } else {
    clearSecureLS()
  }
}

function clearSecureLS () {
  secureLS.clear()
}

function clearForm () {
  const formProfile = $('#form-profile')
  formProfile.reset()
  storeDataInput.checked = false
}

function setCurrentDate (releaseDateInput) {
  const currentDate = new Date()
  releaseDateInput.value = getFormattedDate(currentDate)
}

function showSnackbar (snackbarToShow, showDuration = 6000) {
  snackbarToShow.classList.remove('d-none')
  setTimeout(() => snackbarToShow.classList.add('show'), 100)

  setTimeout(function () {
    snackbarToShow.classList.remove('show')
    setTimeout(() => snackbarToShow.classList.add('d-none'), 500)
  }, showDuration)
}

export function wantDataToBeStored () {
  return storeDataInput.checked
}

export function setReleaseDateTime (releaseDateInput) {
  const loadedDate = new Date()
  releaseDateInput.value = getFormattedDate(loadedDate)
}

export function toAscii (string) {
  if (typeof string !== 'string') {
    throw new Error('Need string')
  }
  const accentsRemoved = removeAccents(string)
  const asciiString = accentsRemoved.replace(/[^\x00-\x7F]/g, '') // eslint-disable-line no-control-regex
  return asciiString
}

export function getProfile (formInputs) {
  const fields = {}
  for (const field of formInputs) {
    let value = field.value
    if (field.id === 'field-datesortie') {
      const dateSortie = field.value.split('-')
      value = `${dateSortie[2]}/${dateSortie[1]}/${dateSortie[0]}`
    }
    if (typeof value === 'string') {
      value = toAscii(value)
    }
    fields[field.id.substring('field-'.length)] = value
  }
  return fields
}

export function getReasons (reasonInputs, context) {
  const reasons = reasonInputs
    .filter(input => input.className.split(' ').includes(context))
    .filter(input => input.checked)
    .map(input => input.value).join(', ')
  return reasons
}

export function prepareInputs (formInputs, reasonInputs, reasonFieldsetsWrapper, reasonAlerts, snackbar, releaseDateInput, contextWrapper) {
  const lsProfile = secureLS.get('profile')

  // Continue to store data if already stored
  storeDataInput.checked = !!lsProfile
  formInputs.forEach((input) => {
    if (input.name && lsProfile && input.name !== 'datesortie' && input.name !== 'heuresortie' && input.name !== 'field-reason') {
      input.value = lsProfile[input.name]
    }
    const exempleElt = input.parentNode.parentNode.querySelector('.exemple')
    if (input.placeholder && exempleElt) {
      input.addEventListener('input', (event) => {
        if (input.value) {
          updateSecureLS(formInputs)
          exempleElt.innerHTML = 'ex.&nbsp;: ' + input.placeholder
        } else {
          exempleElt.innerHTML = ''
        }
      })
    }
  })

  $('#field-birthday').addEventListener('keyup', function (event) {
    event.preventDefault()
    const input = event.target
    const key = event.keyCode || event.charCode
    if (key !== 8 && key !== 46) {
      input.value = addSlash(input.value)
    }
  })

  reasonInputs.forEach(radioInput => {
    radioInput.addEventListener('change', function (event) {
      const isInError = reasonInputs.every(input => !input.checked)
      reasonFieldsetsWrapper.classList.toggle('fieldset-error', isInError)
      reasonAlerts.map(reasonAlert => reasonAlert.classList.toggle('hidden', !isInError))
    })
  })
  $('#cleardata').addEventListener('click', () => {
    clearSecureLS()
    clearForm()
    setCurrentDate(releaseDateInput)
    showSnackbar(clearDataSnackbar, 3000)
  })
  $('#field-storedata').addEventListener('click', () => {
    updateSecureLS(formInputs)
  })
  $('#generate-btn').addEventListener('click', async (event) => {
    event.preventDefault()

    if ($$('.targeted').length === 0) {
      contextWrapper.classList.add('context-wrapper-error')
    }

    const reasons = getReasons(reasonInputs, context)
    if (!reasons) {
      reasonFieldsetsWrapper.classList.add('fieldset-error')
      reasonAlerts.map(reasonAlert => reasonAlert.classList.remove('hidden'))
      // reasonFieldsetsWrapper.scrollIntoView && reasonFieldsetsWrapper.scrollIntoView()
      return
    }

    const invalid = validateAriaFields()
    if (invalid) {
      return
    }
    updateSecureLS(formInputs)
    const pdfBlob = await generatePdf(getProfile(formInputs), reasons, context)

    const creationInstant = new Date()
    const creationDate = creationInstant.toLocaleDateString('fr-CA')
    const creationHour = creationInstant
      .toLocaleTimeString('fr-FR', { hour: '2-digit', minute: '2-digit' })
      .replace(':', '-')

    downloadBlob(pdfBlob, `attestation-${creationDate}_${creationHour}.pdf`)
    showSnackbar(snackbar, 6000)
  })

  const curfewFieldset = $('#curfew-reason-fieldset')
  const quarantineFieldset = $('#quarantine-reason-fieldset')

  $$('.context-button').map(anchor => anchor.addEventListener('click', (event) => {
    contextWrapper.classList.remove('context-wrapper-error')
    reasonFieldsetsWrapper.classList.toggle('hidden', false)
    if (event.target.className.includes('curfew-button')) {
      context = 'curfew'
      curfewFieldset.classList.toggle('targeted', true)
      quarantineFieldset.classList.toggle('targeted', false)
    }
    if (event.target.className.includes('quarantine-button')) {
      context = 'quarantine'
      curfewFieldset.classList.toggle('targeted', false)
      quarantineFieldset.classList.toggle('targeted', true)
    }
  }))
}

export function prepareForm () {
  const formInputs = $$('#form-profile input')
  const snackbar = $('#snackbar')
  const reasonInputs = [...$$('input[name="field-reason"]')]
  const reasonFieldsetsWrapper = $('.fieldset-wrapper')
  const reasonAlerts = $$('.msg-alert')
  const releaseDateInput = $('#field-datesortie')
  const contextWrapper = $('.context-wrapper')
  setReleaseDateTime(releaseDateInput)
  prepareInputs(formInputs, reasonInputs, reasonFieldsetsWrapper, reasonAlerts, snackbar, releaseDateInput, contextWrapper)
}


src/js/pdf-util.js
import { generateQR } from './util'
import { PDFDocument, PageSizes, rgb } from 'pdf-lib'
import fontkit from '@pdf-lib/fontkit'
import imageUrlRF from '../RF.png'
import imageUrlTAC from '../TAC.png'
import LucioleFontBase from '../Luciole-Regular.ttf'
import LucioleBoldFontBase from '../Luciole-Bold.ttf'
import LucioleItalicFontBase from '../Luciole-Regular-Italic.ttf'
import LucioleBoldItalicFontBase from '../Luciole-Bold-Italic.ttf'
import curfewPdfData from '../curfew-pdf-data.json'
import quarantinePdfData from '../quarantine-pdf-data.json'

// const pdfWidth = 595.28
// const pdfHeight = 841.89
// const milimeterWidth = 210
// const milimeterHeight = 297
// const pixelWidth = 892
// const milimeterRatio = 0.35277516462841
const pixelRatio = 1.49845450880258
const sizeRatio = 0.66
export async function generatePdf (profile, reasons, context) {
  let pdfData = curfewPdfData
  if (context === 'quarantine') {
    pdfData = quarantinePdfData
  }
  const creationInstant = new Date()
  const creationDate = creationInstant.toLocaleDateString('fr-FR')
  const creationHour = creationInstant
    .toLocaleTimeString('fr-FR', { hour: '2-digit', minute: '2-digit' })
    .replace(':', 'h')

  const {
    lastname,
    firstname,
    birthday,
    address,
    zipcode,
    city,
    datesortie,
    heuresortie,
  } = profile

  const data = [
    `Cree le: ${creationDate} a ${creationHour}`,
    `Nom: ${lastname}`,
    `Prenom: ${firstname}`,
    `Naissance: ${birthday}`,
    `Adresse: ${address} ${zipcode} ${city}`,
    `Sortie: ${datesortie} a ${heuresortie}`,
    `Motifs: ${reasons}`,
    '', // Pour ajouter un ; aussi au dernier élément
  ].join(';\n')

  const pdfDoc = await PDFDocument.create()
  pdfDoc.registerFontkit(fontkit)
  const page = pdfDoc.addPage(PageSizes.A4)
  const height = page.getSize().height

  // set pdf metadata
  pdfDoc.setTitle('COVID-19 - Déclaration de déplacement')
  pdfDoc.setSubject('Attestation de déplacement dérogatoire')
  pdfDoc.setKeywords([
    'covid19',
    'covid-19',
    'attestation',
    'déclaration',
    'déplacement',
    'officielle',
    'gouvernement',
  ])
  pdfDoc.setProducer('DNUM/SDIT')
  pdfDoc.setCreator('')
  pdfDoc.setAuthor("Ministère de l'intérieur")

  let imageBuffer = await fetch(imageUrlRF).then(res => res.arrayBuffer())
  const imageRF = await pdfDoc.embedPng(imageBuffer)
  imageBuffer = await fetch(imageUrlTAC).then(res => res.arrayBuffer())
  const imageTAC = await pdfDoc.embedPng(imageBuffer)

  let fontBuffer = await fetch(LucioleFontBase).then(res => res.arrayBuffer())
  const fontLuciole = await pdfDoc.embedFont(fontBuffer)
  fontBuffer = await fetch(LucioleBoldFontBase).then(res => res.arrayBuffer())
  const fontLucioleBold = await pdfDoc.embedFont(fontBuffer)
  fontBuffer = await fetch(LucioleItalicFontBase).then(res => res.arrayBuffer())
  const fontLucioleItalic = await pdfDoc.embedFont(fontBuffer)
  fontBuffer = await fetch(LucioleBoldItalicFontBase).then(res => res.arrayBuffer())
  const fontLucioleBoldItalic = await pdfDoc.embedFont(fontBuffer)

  let x = 0
  let size = 0
  let block = 0
  let sameline = false
  let font = fontLuciole

  page.drawImage(imageRF, { x: 30, y: page.getHeight() - 80, width: 56, height: 50 })
  page.drawImage(imageTAC, { x: page.getWidth() - 66, y: page.getHeight() - 80, width: 33, height: 50 })

  let y = height - 75

  pdfData.forEach(item => {
    x = 0
    size = Number((item.size * sizeRatio).toFixed(2)) || 10
    block = (item.block === undefined ? 1 : item.block) // Pas de bloc défini dans le JSON = cofficient interligne 1 = standard
    sameline = (item.sameline === undefined ? false : item.sameline) // Pas de flag défini dans le JSON = on passe à la ligne suivante avant d'écrire (fonctionnement standard)

    switch (item.font) {
      case 'LucioleBold':
        font = fontLucioleBold
        break
      case 'LucioleItalic':
        font = fontLucioleItalic
        break
      case 'LucioleBoldItalic':
        font = fontLucioleBoldItalic
        break
      default:
        font = fontLuciole
    }

    if (item.top) {
      x = item.left / pixelRatio
    }
    const label = item.label || ''
    let text = item.variablesNames ? item.variablesNames.reduce((acc, name) => acc + ' ' + profile[name], label) : label
    const isSelectedReason = reasons.split(', ').includes(item.reason)
    const checkbox = isSelectedReason ? '[x]' : '[ ]'

    if (item.type === 'checkbox') {
      y = y - (block * size)
      const xc = x - 16
      page.drawText(checkbox, { x: xc, y, size, font })
      page.drawText(label, { x, y, size, font })
    }
    if (item.type === 'text') {
      if (!sameline) {
        y = y - (block * size) - 5
      }
      page.drawText(text, { x, y, size, font })
    }
    if (item.type === 'input') {
      if (!sameline) {
        y = y - (block * size) - 5
      }
      text = item.inputs.reduce((acc, cur) => acc + ' ' + profile[cur], text)
      page.drawText(text, { x, y, size, font })
    }
  })

  const qrTitle1 = 'QR-code contenant les informations '
  const qrTitle2 = 'de votre attestation numérique'

  const generatedQR = await generateQR(data)

  const qrImage = await pdfDoc.embedPng(generatedQR)
  const pageX0 = pdfDoc.getPages()[0]
  pageX0.drawText(qrTitle1 + '\n' + qrTitle2, { x: 470, y: 121, size: 6, font, lineHeight: 10, color: rgb(1, 1, 1) })

  pageX0.drawImage(qrImage, {
    x: pageX0.getWidth() - 107,
    y: 21,
    width: 82,
    height: 82,
  })

  pdfDoc.addPage()
  const pageX = pdfDoc.getPages()[1]
  pageX.drawText(qrTitle1 + qrTitle2, { x: 50, y: pageX.getHeight() - 70, size: 11, font, color: rgb(1, 1, 1) })
  pageX.drawImage(qrImage, {
    x: 50,
    y: pageX.getHeight() - 390,
    width: 300,
    height: 300,
  })

  const pdfBytes = await pdfDoc.save()

  return new Blob([pdfBytes], { type: 'application/pdf' })
}
